lms/djangoapps/grades/models.py

```diff
@@ -0,0 +1,229 @@
+    subtree_edited_date = models.DateTimeField('last content edit timestamp')
+    course_version = models.CharField('guid of latest course version', max_length=255)
+
+    # earned/possible refers to the number of points achieved and avaiable to achieve.
+    # graded refers to the subset of all problems that are marked as being graded.
+    earned_all = models.IntegerField()
+    possible_all = models.IntegerField()
+    earned_graded = models.IntegerField()
+    possible_graded = models.IntegerField()
+
+    # track which blocks were visible at the time of grade calculation
+    visible_blocks = models.ForeignKey(VisibleBlocks)
+
+    # use custom manager
+    objects = PersistentSubsectionGradeManager.as_manager()
+
+    def __unicode__(self):
+        """
+        Returns a string representation of this model.
+        """
+        return "PersistentSubsectionGrade user:{}, subsection {}. {}/{} graded, {}/{} all".format(
+            self.user_id,
+            self.usage_key,
+        )
+
+    @classmethod
+    def save_grade(cls, **kwargs):
+        """
+        Wrapper for create_grade or update_grade, depending on which applies.
+        Takes the same arguments as both of thsoe methods.
+        """
+        try:
+            cls.update(**kwargs)
+        except cls.DoesNotExist:
+            cls.objects.create(**kwargs)
+
+    @classmethod
+    def read(cls, user_id, usage_key):
+        """
+        Reads a grade from database
+
+        Arguments:
+            user_id: The user associated with the desired grade
+            usage_key: The location of the subsection associated with the desired grade
+
+        Raises PersistentSubsectionGrade.DoesNotExist if applicable
+        """
+        return cls.objects.get(
+            user_id=user_id,
+            usage_key=usage_key,
+        )
+
+    @classmethod
+    def update(
+        cls,
+        user_id,
+        usage_key,
+        course_version,
+        subtree_edited_date,
+        earned_all,
+        possible_all,
+        earned_graded,
+        possible_graded,
+        visible_blocks,
+    ):
+        """
+        Updates a previously existing grade.
+
+        Requires all the arguments listed in docstring for create_grade
+        """
+        grade = cls.objects.get(
+            user_id=user_id,
+            usage_key=usage_key,
+        )
+
+        # Thanks to repeatable read, there's a non-zero chance of a race condition ocurring on insert.
+        # If that happens, the situation is unrecoverable, as we need to read a new piece of data (a visible_blocks FK)
+        # that won't be visible inside the current transaction. In those cases, log the issue and abort.
+        try:
+            visible_blocks_model = VisibleBlocks.create(blocks=visible_blocks)
+        except IntegrityError:
+            log.error("Race condition hit in robust grading data model. Unrecoverable repeatable-read issue.")
+            raise
```

Comment:
@ormsbee Our handling of the `IntegrityError` is going to have to be a bit different than how it's handled in submissions (with a `pass`). We aren't just inserting a record here, we need the resulting data as a FK before `grade` can be saved. But if we're in repeatable read, we'll never be able to fetch the needed data.

Right now I've just got it raising the error; how else could we handle this?